File: frontend/src/features/widgets/api/widgetsController.ts
import api from "../../../shared/api/api";
import { ApiResponse } from "../../../shared/types/ApiResponse";
import {
    CreateEntriesWidgetDto,
    CreateWidgetDto,
    EntriesWidgetDefinitionDto,
    UpdateEntriesWidgetDto,
    UpdateWidgetDto,
    WidgetDto,
} from "../types/WidgetDto";

export const widgetsController = {
    getWidgets: async (trackerId?: string): Promise<ApiResponse<WidgetDto[]>> => {
        const qs = trackerId ? `?trackerId=${trackerId}` : "";
        return await api.get(`/widgets${qs}`);
    },

    getWidget: async (widgetId: string): Promise<ApiResponse<WidgetDto>> => {
        return await api.get(`/widgets/${widgetId}`);
    },

    createWidget: async (dto: CreateWidgetDto): Promise<ApiResponse<WidgetDto>> => {
        return await api.post("/widgets", dto);
    },

    updateWidget: async (
        widgetId: string,
        dto: UpdateWidgetDto
    ): Promise<ApiResponse<WidgetDto>> => {
        return await api.put(`/widgets/${widgetId}`, dto);
    },

    deleteWidget: async (widgetId: string): Promise<ApiResponse> => {
        return await api.delete(`/widgets/${widgetId}`);
    },

    getEntriesWidgets: async (
        trackerId?: string
    ): Promise<ApiResponse<EntriesWidgetDefinitionDto[]>> => {
        const qs = trackerId ? `?trackerId=${trackerId}` : "";
        return await api.get(`/widgets/entries${qs}`);
    },

    getEntriesWidget: async (
        entriesWidgetId: string
    ): Promise<ApiResponse<EntriesWidgetDefinitionDto>> => {
        return await api.get(`/widgets/entries/${entriesWidgetId}`);
    },

    createEntriesWidget: async (
        dto: CreateEntriesWidgetDto
    ): Promise<ApiResponse<EntriesWidgetDefinitionDto>> => {
        return await api.post("/widgets/entries", dto);
    },

    updateEntriesWidget: async (
        entriesWidgetId: string,
        dto: UpdateEntriesWidgetDto
    ): Promise<ApiResponse<EntriesWidgetDefinitionDto>> => {
        return await api.put(`/widgets/entries/${entriesWidgetId}`, dto);
    },

    deleteEntriesWidget: async (entriesWidgetId: string): Promise<ApiResponse> => {
        return await api.delete(`/widgets/entries/${entriesWidgetId}`);
    },
};

File: backend/src/Operum.API/Controllers/WidgetsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Operum.API.Controllers.Base;
using Operum.Model.DTOs.Widgets.Requests;
using Operum.Service.Interfaces;

namespace Operum.API.Controllers
{
    [Authorize]
    [ApiController]
    [Route("api/[controller]")]
    public class WidgetsController(IWidgetsService widgetsService) : BaseController
    {
        [HttpGet]
        public async Task<IActionResult> GetWidgets([FromQuery] string? trackerId)
        {
            return GetApiResponse(await widgetsService.GetWidgets(trackerId));
        }

        [HttpGet("{widgetId}")]
        public async Task<IActionResult> GetWidget([FromRoute] string widgetId)
        {
            return GetApiResponse(await widgetsService.GetWidget(widgetId));
        }

        [HttpPost]
        public async Task<IActionResult> CreateWidget([FromBody] CreateWidgetDto dto)
        {
            return GetApiResponse(await widgetsService.CreateWidget(dto));
        }

        // Only the widget's name and description -- the definition (result type, code,
        // sources, field mapping) is fixed at creation. See UpdateWidgetDto.
        [HttpPut("{widgetId}")]
        public async Task<IActionResult> UpdateWidget([FromRoute] string widgetId, [FromBody] UpdateWidgetDto dto)
        {
            return GetApiResponse(await widgetsService.UpdateWidget(widgetId, dto));
        }

        // Removes the widget everywhere -- every dashboard placing it loses the placement
        // too (see WidgetsService.DeleteWidget). The client is expected to confirm this
        // with the user before calling it.
        [HttpDelete("{widgetId}")]
        public async Task<IActionResult> DeleteWidget([FromRoute] string widgetId)
        {
            return GetApiResponse(await widgetsService.DeleteWidget(widgetId));
        }

        [HttpGet("entries")]
        public async Task<IActionResult> GetEntriesWidgets([FromQuery] string? trackerId)
        {
            return GetApiResponse(await widgetsService.GetEntriesWidgets(trackerId));
        }

        [HttpGet("entries/{entriesWidgetId}")]
        public async Task<IActionResult> GetEntriesWidget([FromRoute] string entriesWidgetId)
        {
            return GetApiResponse(await widgetsService.GetEntriesWidget(entriesWidgetId));
        }

        [HttpPost("entries")]
        public async Task<IActionResult> CreateEntriesWidget([FromBody] CreateEntriesWidgetDto dto)
        {
            return GetApiResponse(await widgetsService.CreateEntriesWidget(dto));
        }

        [HttpPut("entries/{entriesWidgetId}")]
        public async Task<IActionResult> UpdateEntriesWidget([FromRoute] string entriesWidgetId, [FromBody] UpdateEntriesWidgetDto dto)
        {
            return GetApiResponse(await widgetsService.UpdateEntriesWidget(entriesWidgetId, dto));
        }

        [HttpDelete("entries/{entriesWidgetId}")]
        public async Task<IActionResult> DeleteEntriesWidget([FromRoute] string entriesWidgetId)
        {
            return GetApiResponse(await widgetsService.DeleteEntriesWidget(entriesWidgetId));
        }
    }
}
